File: Core kernel Yaiwes/Backend watchdog workflow adaptativo/sandbox-runtime/iii-sandbox/packages/sdk-python/src/iii_sandbox/stream.py
from __future__ import annotations

import json
from typing import AsyncGenerator

from .client import HttpClient
from .types import LogEvent, SandboxMetrics


class StreamManager:
    def __init__(self, client: HttpClient, sandbox_id: str):
        self._client = client
        self._sandbox_id = sandbox_id

    async def logs(
        self,
        tail: int | None = None,
        follow: bool | None = None,
    ) -> AsyncGenerator[LogEvent, None]:
        params: list[str] = []
        if tail is not None:
            params.append(f"tail={tail}")
        if follow is not None:
            params.append(f"follow={str(follow).lower()}")
        qs = f"?{'&'.join(params)}" if params else ""
        path = f"/sandbox/sandboxes/{self._sandbox_id}/stream/logs{qs}"
        async for line in self._client.stream_get(path):
            try:
                event = LogEvent(**json.loads(line))
                yield event
                if event.type == "end":
                    return
            except (json.JSONDecodeError, ValueError, TypeError, KeyError):
                continue

    async def metrics(
        self, interval: int | None = None
    ) -> AsyncGenerator[SandboxMetrics, None]:
        params: list[str] = []
        if interval is not None:
            params.append(f"interval={interval}")
        qs = f"?{'&'.join(params)}" if params else ""
        path = f"/sandbox/sandboxes/{self._sandbox_id}/stream/metrics{qs}"
        async for line in self._client.stream_get(path):
            try:
                yield SandboxMetrics(**json.loads(line))
            except (json.JSONDecodeError, ValueError, TypeError, KeyError):
                continue

File: Core kernel Yaiwes/Backend watchdog workflow adaptativo/sandbox-runtime/iii-sandbox/packages/sdk-rust/src/stream.rs
use std::pin::Pin;
use std::sync::Arc;

use futures_core::Stream;

use crate::client::HttpClient;
use crate::error::Result;
use crate::types::{LogEvent, SandboxMetrics};

pub struct StreamManager {
    client: Arc<HttpClient>,
    sandbox_id: String,
}

impl StreamManager {
    pub fn new(client: Arc<HttpClient>, sandbox_id: String) -> Self {
        Self { client, sandbox_id }
    }

    pub fn logs(
        &self,
        tail: Option<i64>,
        follow: Option<bool>,
    ) -> Pin<Box<dyn Stream<Item = Result<LogEvent>> + Send + '_>> {
        let mut params = Vec::new();
        if let Some(t) = tail {
            params.push(format!("tail={t}"));
        }
        if let Some(f) = follow {
            params.push(format!("follow={f}"));
        }
        let query = if params.is_empty() {
            String::new()
        } else {
            format!("?{}", params.join("&"))
        };
        let path = format!(
            "/sandbox/sandboxes/{}/stream/logs{}",
            self.sandbox_id, query
        );
        let inner = self.client.stream_get(&path);

        Box::pin(async_stream::stream! {
            use futures_util::StreamExt;
            let mut inner = std::pin::pin!(inner);
            while let Some(item) = inner.next().await {
                match item {
                    Ok(line) => {
                        match serde_json::from_str::<LogEvent>(&line) {
                            Ok(event) => {
                                let is_end = event.event_type == "end";
                                yield Ok(event);
                                if is_end {
                                    return;
                                }
                            }
                            Err(_) => continue,
                        }
                    }
                    Err(e) => {
                        yield Err(e);
                        return;
                    }
                }
            }
        })
    }

    pub fn metrics(
        &self,
        interval: Option<i64>,
    ) -> Pin<Box<dyn Stream<Item = Result<SandboxMetrics>> + Send + '_>> {
        let mut params = Vec::new();
        if let Some(i) = interval {
            params.push(format!("interval={i}"));
        }
        let query = if params.is_empty() {
            String::new()
        } else {
            format!("?{}", params.join("&"))
        };
        let path = format!(
            "/sandbox/sandboxes/{}/stream/metrics{}",
            self.sandbox_id, query
        );
        let inner = self.client.stream_get(&path);

        Box::pin(async_stream::stream! {
            use futures_util::StreamExt;
            let mut inner = std::pin::pin!(inner);
            while let Some(item) = inner.next().await {
                match item {
                    Ok(line) => {
                        match serde_json::from_str::<SandboxMetrics>(&line) {
                            Ok(metrics) => yield Ok(metrics),
                            Err(_) => continue,
                        }
                    }
                    Err(e) => {
                        yield Err(e);
                        return;
                    }
                }
            }
        })
    }
}

fn chrono_now_ms() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as i64
}
